src/Symfony/Component/Security/Csrf/DoubleSubmitCsrfTokenManager.php

```diff
@@ -0,0 +1,247 @@
+        }
+
+        if (!preg_match('/^[-a-zA-Z0-9_]+$/D', $cookieName)) {
+            throw new \InvalidArgumentException('The cookie name contains invalid characters.');
+        }
+    }
+
+    public function getToken(string $tokenId): CsrfToken
+    {
+        return new CsrfToken($tokenId, $this->cookieName);
+    }
+
+    public function refreshToken(string $tokenId): CsrfToken
+    {
+        return new CsrfToken($tokenId, $this->cookieName);
+    }
+
+    public function removeToken(string $tokenId): ?string
+    {
+        return null;
+    }
+
+    public function isTokenValid(CsrfToken $token): bool
+    {
+        if (!$request = $this->requestStack->getCurrentRequest()) {
+            $this->logger?->error('CSRF validation failed: No request found.');
+
+            return false;
+        }
+
+        if (\strlen($token->getValue()) < self::TOKEN_MIN_LENGTH) {
+            $this->logger?->warning('Invalid double-submit CSRF token.');
+
+            return false;
+        }
+
+        if (false === $isValidOrigin = $this->isValidOrigin($request)) {
+            $this->logger?->warning('CSRF validation failed: origin info doesn\'t match.');
+
+            return false;
+        }
+
+        if (true === $isValidDoubleSubmit = $this->isValidDoubleSubmit($request, $token->getValue())) {
+            // Mark the request as validated using double-submit info
+            $request->attributes->set($this->cookieName, 'double-submit');
+
+            return true;
+        }
+
+        if (false === $isValidDoubleSubmit) {
+            return false;
+        }
+
+        // Opportunistically lookup at the session for a previous CSRF validation strategy
+        $session = $request->hasPreviousSession() ? $request->getSession() : null;
+        $usageIndexValue = $session instanceof Session ? $usageIndexReference = &$session->getUsageIndex() : 0;
+        $usageIndexReference = \PHP_INT_MIN;
+        $csrfProtection = $session?->get($this->cookieName);
+        $usageIndexReference = $usageIndexValue;
+
+        // If a previous request was validated using double-submit info, stick to it
+        if ('double-submit' === $csrfProtection && null === $isValidDoubleSubmit) {
+            $this->logger?->warning('CSRF validation failed: double-submit info was used in a previous request but is now missing.');
+
+            return false;
+        }
+
+        // If a previous request was validated using origin info, stick to it
+        if ('origin' === $csrfProtection && null === $isValidOrigin) {
+            $this->logger?->warning('CSRF validation failed: origin info was used in a previous request but is now missing.');
+
+            return false;
+        }
+
+        if (true === $isValidOrigin) {
+            // Mark the request as validated using origin info
+            $request->attributes->set($this->cookieName, 'origin');
+            $this->logger?->debug('CSRF validation accepted using origin info.');
+
+            return true;
```

Comment:
That's on purpose and explained in the PR description: when it works, using Origin is enough to protect against CSRF. For people that don't send the header (because old browser or filtering proxy), the double-submit done in JS works around the missing header. (when neither Origin nor double-submit is found, we reject).

Reply:
To talk about concrete things [example provided on tests](https://github.com/symfony/symfony/pull/58095#discussion_r1741302405) could be used as support.

There is maybe a vocabulary issue here.

What is the signification of "double-submit"?

For me it's about submitting the same string on cookie and on header.

Reply:
[thread resolved] was the last blocker "ish" on me side.